### PR TITLE
Move dependencies to APIBundle

### PR DESCRIPTION
**Pull request by amenophis**

### CODE REVIEW

**Comment by aRn0D** [comment]
@amenophis you need to add those bundles to "suggest" in composer.json (resource bundle)

**Comment by amenophis** [comment]
@aRn0D done, i also add a check before usage of JMS and Hateoas Bundle.
Is it enough ?

**Comment by amenophis** [comment]
@pjedrzejewski @stloyd see other things about this ?

File: src/Sylius/Component/Resource/Exception/MissingBundleException.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Resource\Exception;
+
+/**
+ * @author Jérémy Leherpeur <jeremy@leherpeur.net>
+ */
+class MissingBundleException extends \InvalidArgumentException
```
Comment (aRn0D) [inline src/Sylius/Component/Resource/Exception/MissingBundleException.php]: @pjedrzejewski @stloyd Not sure that it is the right place to put it.

Comment (amenophis) [inline src/Sylius/Component/Resource/Exception/MissingBundleException.php]: @aRn0D why ? Where do i want to put it ?

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 2]
```diff
@@ -461,4 +463,17 @@ protected function isGrantedOr403($permission)
             throw new AccessDeniedHttpException();
         }
     }
+
+    protected function checkAPIBundleRequirements()
```
Comment (stloyd) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: It should be IMO: `validateDependencies()`.

File: src/Sylius/Component/Resource/Exception/MissingBundleException.php (hunk 1 above)
Comment (stloyd) [inline src/Sylius/Component/Resource/Exception/MissingBundleException.php]: IMO this is ok, but it's not invalid argument, it's runtime issue.

Comment (amenophis) [inline src/Sylius/Component/Resource/Exception/MissingBundleException.php]: Changed to RuntimeException

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php (hunk 2 above)
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: Done

**Comment by aRn0D** [comment]
@pjedrzejewski `BazingaHateoasBundle` should be optional, what do you think? @amenophis the serializer must be available for show, index and create method.

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 3]
```diff
@@ -461,4 +463,17 @@ protected function isGrantedOr403($permission)
             throw new AccessDeniedHttpException();
         }
     }
+
+    protected function validateDependencies()
```
Comment (aRn0D) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: I prefer something more generic which will check if a bundle is registered or not.

**Comment by amenophis** [comment]
@aRn0D don't understand your comment about ```BazingaHateoasBundle``` ?

**Comment by amenophis** [comment]
@aRn0D Added a wrapper method in ResourceController for config->isApiRequest() which made the test for needed bundles. Better ?

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 4]
```diff
@@ -461,4 +463,24 @@ protected function isGrantedOr403($permission)
...
+    {
+        if (!$this->config->isApiRequest()) {
+            return false;
+        }
+
+        $bundles = array(
+            'JMSSerializerBundle',
+            'BazingaHateoasBundle'
+        );
+        $enabledBundles = $this->container->getParameter('kernel.bundles');
```
Comment (aRn0D) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: I think we can create a generic method hasDependencies which checks dependencies or throws exception

```php
protected function hasDependencies(array $dependencies)
```

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php [hunk 5]
```diff
@@ -461,4 +463,24 @@ protected function isGrantedOr403($permission)
...
+
+    protected function isApiRequest()
+    {
+        if (!$this->config->isApiRequest()) {
+            return false;
+        }
+
+        $bundles = array(
+            'JMSSerializerBundle',
+            'BazingaHateoasBundle'
```
Comment (aRn0D) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: You are not obliged to use `BazingaHateoasBundle` when you build API it is just better!

**Comment by aRn0D** [comment]
@pjedrzejewski are you agree about what I said ?

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php (hunk 5 above)
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: IIRC Without enable BazingaHateoasBundle for API returns, the returned JSON is OK, but there is never results ...
So seems BazingaHateoasBundle is mandatory for API implementation.

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php (hunk 4 above)
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: I can do that, but don't want to put 'JMSSerializerBundle', 'BazingaHateoasBundle' many times in the code ...

File: src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php (hunk 5 above)
Comment (aRn0D) [inline src/Sylius/Bundle/ResourceBundle/Controller/ResourceController.php]: hummm, you're right !

**Comment by stloyd** [comment]
:-1: for adding yet more code to the controller...

**Comment by pjedrzejewski** [comment]
Actually, my idea is to move the api routing etc. to CoreBundle and rename this bundle to OAuthServer bundle and basically make it integration layer for FOSOAuthServer into Sylius. To make it optional, because you may prefer other authentication methods. What do you think? ApiBundle is not useful I think. All the Sylius api routing/config depends on core so...

**Comment by aRn0D** [comment]
@pjedrzejewski yes you're right  but we can remove `ms/serializer-bundle` and `willdurand/hateoas-bundle` for the `ResourceBundle`, put as suggest. I think we can remove check from the `ResourceController` too.

**Pull request closed without merging**